lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1563,6 +1566,28 @@ def __init__(
         self.framework.observe(events.relation_joined, self._set_scrape_job_spec)
         self.framework.observe(events.relation_changed, self._set_scrape_job_spec)
 
+        if not refresh_event:
+            if len(self._charm.meta.containers) == 1:
+                if "kubernetes" in self._charm.meta.series:
+                    refresh_event = [self._charm.on.update_status]
+                else:
+                    container = list(self._charm.meta.containers.values())[0]
+                    refresh_event = [self._charm.on[container.name.replace("-", "_")].pebble_ready]
+            else:
+                logger.warning(
+                    "%d containers are present in metadata.yaml and "
+                    "refresh_event was not specified. Defaulting to update_status. "
+                    "Metrics IP may not be set in a timely fashion."
+                )
```

Comment:
It is here. We assume a "base" base of "sidecar charms have one container, and that is the default."

For charm authors who use this object, if it doesn't match our baseline assumptions and we don't have another one to fall back on (like "this is podspec with one container"), we are telling the _author_ "this is really important and you should correct it before you ship this."

It's less a warning to Juju admins (who realistically should never see it) and more of an alert to authors saying: "just because you managed to get metrics doesn't mean that it wasn't luck, and you should be explicit."

Reply:
Do we really want to throw exceptions here? That forces a charm author to _immediately_ rebuild and restart their development workflow instead of seeing the warning and fixing it _while_ they continue working.

Reply:
From the thread here it sounds like an exception is exactly what we want:
> "this is really important and you should correct it before you ship this."